### ISSUE TITLE
CoverImage plugin: Remarkable-specific alternative

### ISSUE THREAD

**Issue by gbyl** [koreader/koreader#7506]
As the CoverImage plugin was originally designed with Tolinos in mind, there are a few things that don't work very well with Remarkables:

- The plugin runs automatically on book opening, which can easily double the loading time of files.

- When disabled, on the other hand, the plugin automatically deletes the file generated, which breaks RM1 since it needs a `poweroff.png` file.

- Since the plugin lacks a cache system, it has to generate the image anew every single time.

I don't know how to integrate these changes to the current plugin — a plugin that should perhaps be part of the ScreenSaver menu by @poire-z's [own admission](https://github.com/koreader/koreader/pull/6813#pullrequestreview-514595450). So, I blatantly plagiarized @zwim's work and made a `poweroffimage.koplugin` which addresses the issues listed above while at the same time simplifying the UI to just one item.

_Note: I am sure this is rather sloppy and I have re-invented the wheel in more than one place. So, please share suggestions._

```lua
local Device = require("device")
local Blitbuffer = require("ffi/blitbuffer")
local InfoMessage = require("ui/widget/infomessage")
local UIManager = require("ui/uimanager")
local WidgetContainer = require("ui/widget/container/widgetcontainer")
local RenderImage = require("ui/renderimage")
local ffiutil = require("ffi/util")
local lfs = require("libs/libkoreader-lfs")
local util = require("util")
local _ = require("gettext")

local PowerOffImage = WidgetContainer:new{
    name = "poweroffimage",
    is_doc_only = true,
}

local function getFileName(fullpath)
    local path, name_extension = util.splitFilePathName(fullpath)
    local name, extension = util.splitFileNameSuffix(name_extension)
    return name:gsub(" ","_"):lower()
end

function PowerOffImage:init()
    self.poweroff_image_path = "/usr/share/remarkable/poweroff.png"
    self.poweroff_image_cache = "/home/root/.coverimages/"..getFileName(self.ui.document.file)..".png"
    self.ui.menu:registerToMainMenu(self)
end

function PowerOffImage:createPowerOffImage(doc_settings)
    if not util.pathExists(self.poweroff_image_cache) then
        local poweroff_image = self.ui.document:getCoverPageImage()
        local s_w, s_h = Device.screen:getWidth(), Device.screen:getHeight()
        local i_w, i_h = poweroff_image:getWidth(), poweroff_image:getHeight()
        local scale_factor = math.min(s_w / i_w, s_h / i_h)

        local scaled_w, scaled_h = math.floor(i_w * scale_factor), math.floor(i_h * scale_factor)
        poweroff_image = RenderImage:scaleBlitBuffer(poweroff_image, scaled_w, scaled_h)
        local image = Blitbuffer.new(s_w, s_h, poweroff_image:getType())
        if s_w > scaled_w then
            image:blitFrom(poweroff_image, math.floor((s_w - scaled_w) / 2), 0, 0, 0, scaled_w, scaled_h)
        else
            image:blitFrom(poweroff_image, 0, math.floor((s_h - scaled_h) / 2), 0, 0, scaled_w, scaled_h)
        end

        poweroff_image:free()

        if not image:writeToFile(self.poweroff_image_cache, "png", 75) then
            UIManager:show(InfoMessage:new{
                text = _("Error writing file").."\n"..self.poweroff_image_cache,
                show_icon = true,
            })
        else
            ffiutil.copyFile(self.poweroff_image_cache, self.poweroff_image_path)
            UIManager:show(InfoMessage:new{
                text = _("Cover image written to:").."\n"..self.poweroff_image_path,
                show_icon = true,
            })
        end
    
        image:free()
    else
        ffiutil.copyFile(self.poweroff_image_cache, self.poweroff_image_path)
        UIManager:show(InfoMessage:new{
            text = _("Cover image copied to:").."\n"..self.poweroff_image_path,
            show_icon = true,
        })
    end
end

function PowerOffImage:addToMainMenu(menu_items)
    menu_items.poweroffimage = {
        sorting_hint = "screen",
        text = _("Set book cover as poweroff image"),
        callback = function()
            self:createPowerOffImage(self.ui.doc_settings)
        end,
    }
end

return PowerOffImage
```

**Comment by pazos** [comment]
The plugin looks fine as an user plugin, but IMHO doesn't make sense upstream.

What we need is a person with a ReMarkable and some time to circunvent the current limitations of the system. As I said elsewhere there's no need at all to user the coverimage plugin on the remarkable. Is a NTX board like kobos or cervantes. We run as root and have full control over the device. What we need to do is implement our own `suspend`, `reboot` and `poweroff` routines instead of using the `systemd` services used by the system. Then we can use the full blown screen saver, including all its options.

The `suspend` routine could be a bit tricky to get it right but the poweroff/reboot routines should be a matter of replacing https://github.com/koreader/koreader/blob/master/frontend/device/remarkable/device.lua#L189 and https://github.com/koreader/koreader/blob/master/frontend/device/remarkable/device.lua#L193 with `poweroff`/`halt` and `reboot`, directly. 

For the suspend routine see my comment in https://github.com/koreader/koreader/issues/6028#issuecomment-610411019

**Comment by gbyl** [comment]
> The plugin looks fine as an user plugin, but IMHO doesn't make sense upstream.

I thoroughly agree that this makes no sense upstream. I decided to submit it here merely because there are certain things explained better with code than with conventional words.

**Comment by gbyl** [comment]
> but the poweroff/reboot routines should be a matter of replacing https://github.com/koreader/koreader/blob/master/frontend/device/remarkable/device.lua#L189 and https://github.com/koreader/koreader/blob/master/frontend/device/remarkable/device.lua#L193 with poweroff/halt and reboot, directly.

@pazos Unfortunately, I am seeing exactly the same behavior with `os.execute("systemctl poweroff")` and `os.execute("poweroff")` and `os.execute("/sbin/poweroff")`. The Screensaver is drawn briefly and then system poweroff image is displayed. As described on https://github.com/koreader/koreader/issues/7495

**Comment by zwim** [comment]
Concerning your last point: I currently am working on a cache mechanism, so loading time should get faster on consecutive book openings.

For your second point: That's what the fallback thing is for. Set your favorite picture with `Set fallback image path` and enable `Turn on fallback image`. Then you have either the book cover or your picture stored.

**Comment by Frenzie** [comment]
Regarding cache, maybe the global cache could be worthwhile?

**Comment by gbyl** [comment]
@zwim That's excellent news! I am looking forward to testing the new version.

And yes, the fallback functionality does solve that issue. But I found that `Turn on fallback image` and `Save cover image` can conflict with each other. Particularly, the fallback was not being copied in place after toggling `Turn on fallback image` on/off — somehow the `os.remove` always taking precendence over the `ffiutil.copyFile`.

**Comment by NiLuJe** [comment]
PNG relies more heavily on "simple" (and lossless) RLE-like encoding tricks than JPG. The less color changes, the smaller the PNG. A JPG, on the other hand, will (among a host of other things) quantize to limit those abrupt color changes in the first place. And do all kinds of other lossy trickery to cadge a few bytes here and there.

TL;DR: On eInk, PNGs are worth it if you pre-process the source correctly (e.g., dithered to 16c), and actually encode it as such (i.e., palletted). Otherwise, the size can and will balloon with most content, unlike a JPG.

**Comment by gbyl** [comment]
@NiLuJe I checked back the files with most input/output discrepancy and it was exactly what you would expect from your explanation! The covers with more colors increase exponentially in size.

Thankfully, even if we don't do any pre-processing, the cache mechanism that's in the works might just be enough to speed things up. It certainly worked wonders in my pseudo-plugin.

Side Note: This might also explain why CREngine clamps the number of colors (resulting at times in lost gradients/posterized look). The same image in the Screensaver and in CREngine can look dramatically different. Disabling such compression would probably make for excessively long loading times and page turns.

**Comment by zwim** [comment]
> And yes, the fallback functionality does solve that issue. But I found that `Turn on fallback image` and `Save cover image` can conflict with each other. Particularly, the fallback was not being copied in place after toggling `Turn on fallback image` on/off — somehow the `os.remove` always taking precendence over the `ffiutil.copyFile`.

The fallback image gets copied `onCloseDocument`.  Just toggling the menu has no instant effect.

**Comment by zwim** [comment]
> Regarding cache, maybe the global cache could be worthwhile?

If you mean the `koreader/cache` directory I am fine with it. 
But if you mean some other cache we do have the problem that the cover image may be streched, has some black/white/gray backround and maybe some format (png,jpg,bmp).

**Comment by poire-z** [comment]
> Side Note: This might also explain why CREngine clamps the number of colors (resulting at times in lost gradients/posterized look). The same image in the Screensaver and in CREngine can look dramatically different. Disabling such compression would probably make for excessively long loading times and page turns.

Dunno how you have those set, but 2 options might change this with crengine:
- Screen > [ ] Color rendering
- Bottom menu, right most panel, Image scaling: fast | best (@NiLuJe made sure we could be the bestest :)

**Comment by Frenzie** [comment]
@zwim 
> But if you mean some other cache we do have the problem that the cover image may be streched, has some black/white/gray backround and maybe some format (png,jpg,bmp).

I meant https://github.com/koreader/koreader/blob/17e6049e3004bafbf7897240ba123e1408516222/frontend/cache.lua if it makes sense for the purpose. (Possibly not since you might want this cache to live longer.) I'm not sure what the cover image being stretched etc. has to do with anything though. Those are simply properties you'd include in the cache key.

**Comment by zwim** [comment]
@Frenzie 
The stretched etc. things were meant, that we have to create the image in CoverImage (so it is NOT ready in an existing cache).

**Comment by NiLuJe** [comment]
@gbyl: As @poire-z said, unless you disabled dithering (in dev opts) (or enabled color rendering but that shouldn't be exposed in the UI on an rM), you should *never* see any posterization artifacts in CRe (barring *extreme* gradients).

**Comment by NiLuJe** [comment]
Oh, wait, on rM, if your launcher is inhibiting the framebuffer depth switch, you *will* lose dithering. I don't quite recall if that still applies nowadays, because I petitioned for them not to break that ;p.

**Comment by gbyl** [comment]
@poire-z @NiLuJe 

> Screen > [ ] Color rendering

This option is not available, but just in case I added manually `["color_rendering"] = false` to my settings. Also, I set the image scaling to best. The rendering discrepancy remains nonetheless.

Take for example the epub attached below. On CREngine it shows as one gray rectangle. On Screensaver you can actually see the shapes.

[screensaver-vs-crengine-imagerendering.zip](https://github.com/koreader/koreader/files/6260095/screensaver-vs-crengine-imagerendering.zip)

**Comment by NiLuJe** [comment]
Oh, that might be another matter entirely (in that CRe's grayscaling code takes some *extreme* shortcuts for performance, while we use the gamma-correct equation ^^).

I'll double-check later tonight ;).

**Comment by Frenzie** [comment]
@zwim

> The stretched etc. things were meant, that we have to create the image in CoverImage (so it is NOT ready in an existing cache).

I'm talking about cache *systems*.

**Comment by NiLuJe** [comment]
> @poire-z @NiLuJe
> 
> > Screen > [ ] Color rendering
> 
> This option is not available, but just in case I added manually `["color_rendering"] = false` to my settings. Also, I set the image scaling to best. The rendering discrepancy remains nonetheless.
> 
> Take for example the epub attached below. On CREngine it shows as one gray rectangle. On Screensaver you can actually see the shapes.
> 
> [screensaver-vs-crengine-imagerendering.zip](https://github.com/koreader/koreader/files/6260095/screensaver-vs-crengine-imagerendering.zip)

Yep, that's a fidelity loss in CRe's grayscaling pass ;).

**Comment by gbyl** [comment]
Issues addressed by https://github.com/koreader/koreader/pull/7510
